optimization/ht_functions.py

```diff
@@ -0,0 +1,385 @@
+        """
+        res = minimize_scalar(_error, bounds=(1, 1e9), args=(self),
+        method='Bounded', options={'xatol':1e-5})
+        # round up to nearest N_channel
+        self.guess = math.ceil(self.guess)
+        self.N_channels = self.guess
+        # use rounded N_channel to calculate flow characteristics
+        self.set_geom()
+        self.characterize_flow()
+        self.get_q_per_channel()
+
+    def calc_reactor_mass(self):
+        """Based on results of the iteration, calculate the reactor mass.
+        
+        Modified Attributes:
+        --------------------
+            Vol_fuel: total fuel volume [m^3]
+            mass: total fuel mass[kg]
+        """
+        self.Vol_fuel = self.A_fuel * self.L * self.N_channels
+        self.mass = self.Vol_fuel * rho_fuel
+
+class ParametricSweep():
+    """Class to store results of parametric sweeps for 1D flow channel analysis.
+    
+    """
+    titles = {'mass' : ("Total Fuel Mass", "m [kg]"),
+              'Re' : ("Reynolds Number", "Re [-]"),
+              'G_dot' : ("Reactor Mass Flux", "G_dot [kg/m^2-s]"),
+              'N_channels' : ("Number of Fuel Channels", "N Channels [-]"),
+              'Nu' : ("Nusselt Number", "Nu [-]"),
+              'dp' : ("Subchannel Pressure Drop", "dP [Pa]"),
+              'h_bar' : ("Heat Transfer Coefficient", "h [W / m^2 - K]"),
+              'q_per_channel' : ("Total Subchannel Generation", "q/channel [W]"),
+              'q_bar' : ("Average Volumetric Generation", "q_bar [W/m^3]"),
+              'v' : ("Flow Velocity", "v [m/s]"),
+              'R_fuel' : ("Resistance to Conduction in Fuel", "R_fuel [K/W]"),
+              'R_clad' : ("Resistance to Conduction in Clad", "R_clad [K/W]"),
+              'R_conv' : ("Resistance to Convection", "R_conv [K/W]"),
+              'R_tot' : ("Total Resistance to Heat Transfer", "R_tot [K/W]"),
+              'AR' : ("Approximate Core Aspect Ratio", "AR [-]")
+             }
+
+    D = 0; PD = 0;
+    
+    # dict to save data for plotting
+    data = {k: [] for k in titles.keys()}
+    min_idx = None; min_jdk = None; min_mass = 0; minD = 0; min_PD = 0;
+    select_AR = False
+    
+    def __init__(self, D, PD, N, select_AR):
+        self.D = D
+        self.PD = PD
+        self.N = N
+        self.select_AR = select_AR
+        for key in self.data:
+            self.data[key] = np.empty([N,N])
+
+    def save_iteration(self, iteration, i, j):
+        """ Save the data from each iteration of the parametric sweep. 
+        """
+        for key in self.data.keys():
+            self.data[key][i][j] = iteration.__dict__[key]
+
+    def get_min_data(self):
+        """ After the parametric sweep is complete, find the minimum calculated
+        fuel mass corresponding with a valid aspect ratio. Save the flowdata for that
+        point, display the PD, D and mass for the optimized configuration.
+        """
+        
+        if False not in np.isnan(self.data['mass']):
```

Comment:
I now realize that this is deprecated. Before I had check_dp(), if a configuration violated the pressure limit, I would set the mass values to NaN, then I would eliminate those from the minimum mass consideration. The False not in ensured that at least ONE configuration would meet the dP requirements. Now (because of check_dp() ) every configuration should meet the requirements. I am going to rehaul this method.